### PR TITLE
Harden subagent review follow-ups (Fixes #2472)

### LINKED ISSUE #2472: Remediate OCR review findings from PR #2415 (subagent runtime isolation)
## Summary

PR #2415 (Fixes #2410 — subagent runtime isolation) generated 50 OCR review threads across 3 automated review runs. After evaluation against the actual committed source code, the findings break down as follows:

- **~15 false positives** (OCR misread the code structure)
- **~8 legitimate small fixes** (missing JSDoc, test naming, log path improvement, missing test edge cases)
- **~15 low-priority / intentional patterns** (valid concerns that are already addressed or are deliberate design choices)
- **~7 design/architecture concerns** (broader improvements beyond the PR's scope)

### Tier 1 — Small Code Fixes (do in 0.10.0)

1. **Add JSDoc to `markToolResponsesRecorded`** (TurnProcessor.ts:347, StreamProcessor.ts:124) — two methods lack documentation
2. **Add explicit `public` modifier** to `markToolResponsesRecorded` in StreamProcessor.ts
3. **Log `resolvedPath` instead of `expandedKeyfile`** in subagentSettingsPopulation.ts keyfile error handler — `path.resolve()` may change the displayed path
4. **Add `(Issue #2410)` reference** to the one test in subagentToolProcessing.test.ts that's missing it
5. **Add non-empty string test case** to AnthropicMessageValidator.stripEmptyText.test.ts ("leaves string content untouched" only tests empty string)
6. **Add edge case tests** to AnthropicMessageValidator.stripEmptyText.test.ts (empty content arrays, non-text-only blocks, Unicode whitespace)
7. **Rename** AnthropicMessageValidator.stripEmptyText.test.ts to match the function name `stripEmptyTextBlocks`

### Tier 2 — Test Hygiene (do in 0.10.0)

8. **Remove unused `Type` mock** from GeminiProvider.auth.test.ts if confirmed unnecessary for module loading
9. **Document or remove unused mock methods** (`set`, `updateSettings`, `getAllGlobalSettings`) in GeminiProvider.auth.test.ts mockSettingsService
10. **Add comment explaining why private property mutation** is used in nonInteractiveCli.slashCommandsAndThinking.test.ts (no public API seam exists)

### Tier 3 — Design Improvements (evaluate for 0.10.0 or 0.11.0)

11. **Secrets handling in SettingsService** — evaluate whether auth-key values should be flagged as sensitive and excluded from diagnostics/event payloads (PnavJ)
12. **structuredClone failure UX** — consider whether non-cloneable ephemeral warnings should be escalated beyond debug-level (PnavL)
13. **getSetting null guard** — add defensive null check in subagentSettingsAccess.ts getSetting helper (PrS2o)
14. **Strict null-checking consistency** — unify truthy vs strict-undefined checks in GeminiProvider Vertex auth helpers (Pnavt)

## Context

All 50 OCR threads on PR #2415 are being resolved with explanatory comments. This issue tracks the actionable items that should be addressed as follow-up work rather than blocking the PR.

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Remediates the actionable, issue-scoped follow-ups from the subagent runtime isolation review. This hardens auth-key diagnostics boundaries, resolves path/error and null-safety edge cases, improves Anthropic empty-content handling, and cleans up the related tests without expanding into generalized secret serialization.

## Dive Deeper

- documents eager tool-response tracking and the unavoidable private-state test seam
- reports the fully resolved auth keyfile path for missing and empty files
- makes subagent setting access null-safe and keeps Vertex auth completeness checks consistent for blank values
- renames the Anthropic strip-empty-text test to match the function and covers non-empty strings, Unicode whitespace, true empty arrays, and non-text-only blocks
- removes unused Gemini auth-test mocks and keeps tests on public client-creation behavior
- marks only auth-key and its existing aliases as sensitive in registry metadata
- redacts auth-key values at SettingsService diagnostics/change-event and CLI diagnostics boundaries while preserving profile export values
- deliberately does not add recursive serialization, cycle handling, tool-key sensitivity, keyfile sensitivity, or new warning-channel architecture
- leaves StreamProcessor implicit-public because the repository ESLint configuration forbids explicit public accessibility
- retains the existing structured-clone warning layer because this synchronous settings-copy code has no established user-facing warning channel

## Reviewer Test Plan

1. Run the focused issue tests:

       npx vitest run packages/settings/src/__tests__/SettingsService.test.ts packages/settings/src/__tests__/settingsRegistry.redaction.test.ts packages/cli/src/ui/commands/diagnosticsCommand.edges.spec.ts packages/providers/src/anthropic/AnthropicMessageValidator.stripEmptyTextBlocks.test.ts packages/providers/src/gemini/GeminiProvider.auth.test.ts packages/agents/src/core/subagentSettingsAccess.test.ts packages/agents/src/core/subagentSettingsPopulation.test.ts packages/agents/src/core/subagentToolProcessing.test.ts packages/cli/src/nonInteractiveCli.slashCommandsAndThinking.test.ts

2. Confirm diagnostics replace auth-key/apiKey/api-key values with [REDACTED], while auth-keyfile and profile export remain unchanged.
3. Confirm relative missing/empty keyfile warnings show the absolute path actually used for I/O.
4. Confirm whitespace-only Vertex project/location values fail configuration and Anthropic empty content arrays receive the existing placeholder.
5. Run full verification:

       npm run test
       npm run lint
       npm run typecheck
       npm run format
       npm run build
       bun scripts/start.ts --profile-load ollamakimi "write me a haiku and nothing else"

All commands above passed on macOS. Open Code Review inspected all 20 changed files; its actionable strict-equality and assertion-strength findings were addressed.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ✅  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Fixes #2472

### CODE REVIEW

File: packages/cli/src/ui/commands/diagnosticsCommand.ts [hunk 1]
```diff
@@ -302,31 +294,26 @@ function appendEphemeralSettings(
   ephemeralSettings: Record<string, unknown>,
   logger: DebugLogger,
 ): void {
+  const redactedSettings = redactSensitiveValues(ephemeralSettings);
   diagnostics.push('\n## Ephemeral Settings');
   logger.debug(
     () =>
-      `[diagnostics] ephemeral settings ${JSON.stringify(ephemeralSettings)}`,
+      `[diagnostics] ephemeral settings ${JSON.stringify(redactedSettings)}`,
   );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/ui/commands/diagnosticsCommand.ts:302]: No change needed. The logger and rendered diagnostics both use the same redactedSettings object, whose auth-key values are already replaced with [REDACTED]. Raw ephemeral settings are not logged.

File: packages/agents/src/core/subagentSettingsPopulation.test.ts [hunk 2]
```diff
@@ -160,6 +160,67 @@ describe('populatePreActivationSettings and populatePostActivationSettings', ()
...
+      );
+
+      expect(warn).toHaveBeenCalledExactlyOnceWith(
+        expect.stringContaining(emptyKeyfile),
+      );
+    } finally {
+      warn.mockRestore();
+      fs.rmSync(tempDir, { recursive: true, force: true });
+    }
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/subagentSettingsPopulation.test.ts:222]: No change. These are two explicit behavioral branches—missing file and empty file—with distinct setup. A helper for only two cases would hide the lifecycle under test and add abstraction outside this focused remediation.

[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/subagentSettingsPopulation.test.ts:222]: No change. This duplicates the preceding suggestion. The two tests intentionally keep their missing-file and empty-file setup visible; extracting a helper for two branches would not improve this issue-scoped test.

File: packages/agents/src/core/subagentSettingsPopulation.test.ts [hunk 3]
```diff
@@ -160,6 +160,67 @@ describe('populatePreActivationSettings and populatePostActivationSettings', ()
...
+      populatePostActivationSettings(
+        service,
+        profile,
+        'relative-missing-keyfile-profile',
+        new Set(),
+      );
+
+      expect(warn).toHaveBeenCalledExactlyOnceWith(
+        expect.stringContaining(missingKeyfile),
+      );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/subagentSettingsPopulation.test.ts:186]: No change. Logging the resolved path is an explicit requirement of issue 2472 so diagnostics identify the path actually used for I/O. auth-keyfile is path metadata, not an authentication secret, and is intentionally outside the auth-key redaction scope.

File: packages/settings/src/settings/SettingsService.ts [hunk 4]
```diff
@@ -44,6 +53,23 @@ type SettingsEventListener =
...
 
+/**
+ * Snapshot returned by {@link SettingsService.getDiagnosticsData}.
+ * Sensitive setting values (e.g. `auth-key`) are replaced with
+ * `[REDACTED]`; all other values are preserved verbatim.
+ */
+export interface DiagnosticsData {
+  provider: string;
+  model: string;
+  profile: string | null;
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/settings/src/settings/SettingsService.ts:64]: No change. The current fallback is always the non-empty string openai, so activeProvider is a string. Repository typecheck and all platform CI typechecks pass; the null fallback described here is not present in this implementation.

File: packages/settings/src/settings/SettingsService.ts [hunk 5]
```diff
@@ -80,8 +106,8 @@ export class SettingsService extends EventEmitter {
 
     this.eventEmitter.emit('change', {
       key,
-      oldValue,
-      newValue: value,
+      oldValue: redactEventValue(key, oldValue),
+      newValue: redactEventValue(key, value),
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/settings/src/settings/SettingsService.ts:110]: Audited. The production settings-event consumers use only key and provider metadata for cache invalidation; they do not consume oldValue or newValue. Redaction at the event boundary is the intended issue-scoped security behavior and integration coverage passes.

File: packages/settings/src/settings/SettingsService.ts [hunk 6]
```diff
@@ -44,6 +53,23 @@ type SettingsEventListener =
   | (() => void)
   | ((...args: unknown[]) => void);
 
+/**
+ * Snapshot returned by {@link SettingsService.getDiagnosticsData}.
+ * Sensitive setting values (e.g. `auth-key`) are replaced with
+ * `[REDACTED]`; all other values are preserved verbatim.
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/settings/src/settings/SettingsService.ts:59]: No change. [REDACTED] is the intentional public diagnostics contract, while REDACTED_VALUE is its implementation constant. Tests pin the contract, so the documentation is accurate rather than an independently drifting implementation detail.

File: packages/agents/src/core/subagentSettingsPopulation.ts [hunk 7]
```diff
@@ -198,8 +198,8 @@ function tryLoadApiKeyFromKeyfile(
   expandedKeyfile: string,
   service: SettingsService,
 ): void {
+  const resolvedPath = path.resolve(expandedKeyfile);
   try {
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/subagentSettingsPopulation.ts:202]: No change. expandedKeyfile is statically a string, so path.resolve does not throw for a valid call. Keeping resolvedPath in catch scope is required to report the actual attempted path; adding broader malformed-runtime handling is outside this narrow issue.

File: packages/agents/src/core/subagentSettingsPopulation.ts [hunk 8]
```diff
@@ -351,8 +351,9 @@ function populateGeneralEphemerals(
 ): void {
   const ephemerals = profile.ephemeralSettings as
     | Record<string, unknown>
+    | null
     | undefined;
-  if (ephemerals === undefined) {
+  if (ephemerals === null || ephemerals === undefined) {
     return;
   }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/subagentSettingsPopulation.ts:358]: Thanks. The explicit null and undefined checks are intentional and covered by the focused settings-access tests.

File: packages/cli/src/ui/commands/diagnosticsCommand.ts [hunk 9]
```diff
@@ -302,31 +294,26 @@ function appendEphemeralSettings(
   ephemeralSettings: Record<string, unknown>,
   logger: DebugLogger,
 ): void {
+  const redactedSettings = redactSensitiveValues(ephemeralSettings);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/ui/commands/diagnosticsCommand.ts:297]: No change; this is outside the approved scope. Settings diagnostics operate on shallow registry records, where auth-key and its aliases are actual setting keys. Recursive traversal would create a generalized serializer with new array, cycle, and arbitrary-payload semantics that issue 2472 does not request.

File: packages/agents/src/core/subagentSettingsPopulation.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/subagentSettingsPopulation.test.ts:222]: No change. This repeats the earlier duplication suggestion. The two explicit cases keep setup and cleanup local to each behavioral branch; introducing a parameterized abstraction for two cases would expand rather than simplify this focused patch.

**Comment by acoliver** [comment]
Compatibility note for the LLxprt review: getDiagnosticsData still returns the same object fields and every production consumer was audited. Existing callers consume model, provider, profile, and settings fields directly, and the Node consumer smoke plus full TypeScript matrix pass. The named DiagnosticsData type narrows and documents the established shape; it does not change runtime structure. Raw auth-key diagnostics are intentionally no longer part of the contract. No additional compatibility abstraction was added because that would expand this focused security remediation.
